PEP 11: define tiered platform support (#2442)
python/peps: pep-0011.txt

unnecessary as the platforms evolve as well.

Allowing these fragments to grow poses the risk of
unmaintainability: without having experts for a large number of
platforms, it is not possible to determine whether a certain
change to the CPython source code will work on all supported
platforms.

To reduce this risk, this PEP specifies what is required for a
platform to be considered supported by CPython as well as providing a
procedure to remove code for platforms with few or no CPython
users.

This PEP also lists what plaforms *are* supported by the CPython
interpreter. This lets people know what platforms are directly
supported by the CPython development team.

Support tiers
=============

Platform support is broken down into *tiers*. Each tier comes with
different requirements which lead to different promises being made
about support.

Tier 1
------

- `CI failures <https://github.com/python/cpython/actions/workflows/build.yml?query=branch%3Amain+is%3Acompleted>`__ block releases.
- Changes which would break the ``main`` branch are not allowed to be merged;
  any breakage may be reverted immediately.
- All core developers are responsible to keep these platforms,

>   and thus ``main``, working.

Reviewer: This rules doesn't work. "Everyone is reponsible" means in practice "nobody is responsible". IMO you must also put two names in front of each platform.

You can put on ``x86_64-\*-linux-gnu    glibc, gcc``.

Author: This is being discussed over on python-committers.

- Promotion to this tier requires team consensus/SC approval.

> =================== =====
> Target Triple       Notes
> =================== =====
> i686-windows-msvc
> x86_64-windows-msvc
> x86_64-apple-darwin BSD libc, clang
> x86_64-linux-gnu    glibc, gcc
> =================== =====

Reviewer 2: I would prefer to use long names here and indicate any vendor with either ``*`` or ``unknown``.

```suggestion
====================== =====
Target Triple          Notes
====================== =====
i686-pc-windows-msvc
x86_64-pc-windows-msvc
x86_64-apple-darwin    BSD libc, clang
x86_64-*-linux-gnu     glibc, gcc
====================== =====
```

Tier 2
------

- Must have a reliable buildbot.
- At least **two** core developers are signed up to support the platform.
- Changes which break any of these platforms are to be fixed or
  reverted within 24 hours.
- Failures on these platforms block a release.
- Promotion to this tier requires consensus/SC approval.

> 

Reviewer: I'm not sure that buildbot URLs are reliable. In the past, some ids changed. Moreover, an URL is not easy to read.

Can you either replace the URL with the buildbot name, or add a label to the URL?

For example, replace ``https://buildbot.python.org/all/#/builders/125`` with ``aarch64 Fedora Stable 3.x``.

Reviewer 3: Or better yet (and per PEP 12),

```rst
`aarch64 Fedora Stable 3.x <https://buildbot.python.org/all/#/builders/125>`__
```

Author: @CAM-Gerlach that suggestion doesn't work since Victor's worry is about URL stability.

@vstinner what I can do is once this update goes in have a `tier-2` label added to the appropriate buildbots and then simply link to that label.

Reviewer 3: I understand that; I thought by "add a label to the URL" he meant using a link with the buildbot name as descriptive link text, `aarch64 Fedora Stable 3.x`, so it could still be located if its URL changes, but I'm guessing its referring to something else with the buildbots I'm unfamiliar with?

Author: He's referring to the labels you can apply to individual buildbots and then pull up a list of them, e.g. all of the 3.10 buildbots can be found with the `3.10` label at https://buildbot.python.org/all/#/builders?tags=%2B3.10 .

Reviewer 3: Gotcha, thanks :+1:

====================== ========================== ============================================== ========
Target Triple          Notes                      Buildbot                                       Contacts
====================== ========================== ============================================== ========

> aarch64-apple-darwin   clang                      https://buildbot.python.org/all/#/builders/725 XXX, XXX

Reviewer 4: If `x86_64-apple-darwin` is a Tier 1 platform, then this must be as well. As of today, nearly all Macs being manufactured by Apple are of this architecture (i.e. Apple Silicon) and Apple has made clear that the few remaining legacy Intel-based Macs still being offered will be replaced in the near future (likely by the end of 2022). So, eventually `x86_64-apple-darwin` will fade in importance as older Macs are retired but that will take many years.

Reviewer 5: Until this is available in Github CI so it can be a reliable **merge** blocker I don't believe we have a choice but to list it as Tier 2.  The tier shouldn't be interpreted as importance to the world.  Anything listed in 1 or 2 is already a **release** blocker.

Reviewer 5: _(and agreed, x86_64 and universal Darwin binaries will likely remain important through 2028)_

Reviewer 4: OK, if the lack of GitHub CI support is the hangup at the moment, I agree with that assessment although we should perhaps note that, once it is available, we need to move this to Tier 1.  You can add my name and I assume @ronaldoussoren 's name. Perhaps also @ambv ?

Reviewer 6: You can add my name as well.

Reviewer 7: You can add my name too cc @ronaldoussoren @ned-deily 
My main equipment is m1 MacBook pro now and my whole core dev's experience is based on apple x86 and aarch64, so I think that I am ready to support this platform :)

Author: ```suggestion
aarch64-apple-darwin   clang                      https://buildbot.python.org/all/#/builders/725 Ned Deily, Ronald Oussoren, Dong-he Na
```

@ned-deily yes, the assumption is as soon as we have CI support for a platform and we are using it the platform will get promoted to tier 1.

> aarch64-linux-gnu      glibc, gcc                 https://buildbot.python.org/all/#/builders/125 XXX

Reviewer 8: ```suggestion
aarch64-linux-gnu      glibc, gcc                 https://buildbot.python.org/all/#/builders/125 Petr Viktorin
```

Reviewer: You can add me: Victor Stinner

>                        glibc, clang               https://buildbot.python.org/all/#/builders/234 Victor Stinner, XXX

Reviewer 5: You add my name here for `aarch64-linux-gnu glibc, clang`

Author: ```suggestion
                       glibc, clang               https://buildbot.python.org/all/#/builders/234 Victor Stinner, Gregory P. Smith
```

>                        glibc, clang               https://buildbot.python.org/all/#/builders/234 Victor Stinner, XXX

Author: ```suggestion
                       glibc, clang               https://buildbot.python.org/all/#/builders/234 Victor Stinner, Gregory P. Smith
```

> aarch64-windows-msvc                              https://buildbot.python.org/all/#/builders/729 XXX, XXX

Reviewer 9: ```suggestion
aarch64-windows-msvc                              https://buildbot.python.org/all/#/builders/729 Steve Dower, XXX
```

Not sure who else to put for this one - I assume once VMs are available one of the people running existing buildbots will happily add one, but with only physical hardware right now it's a small group.

Author: Yeah, I was afraid this might be currently limited until we have CI hardware or more folks have the hardware. Or as long as someone else is will to stay on top of the buildbot and diagnose issues from it we could put them here.

Reviewer 9: Niyas Sait (@nsait-linaro) owns the current buildbot. If he's happy to be listed for now, that's fine by me

Author: But Niyas isn't a core dev, so they can't be listed.

> powerpcle-linux-gnu    glibc, gcc                 https://buildbot.python.org/all/#/builders/90  XXX

Reviewer 8: ```suggestion
powerpcle-linux-gnu    glibc, gcc                 https://buildbot.python.org/all/#/builders/90  Petr Viktorin
```

Reviewer: You can add me: Victor Stinner

>                        glibc, clang               https://buildbot.python.org/all/#/builders/435 XXX

Reviewer: You can add me: Victor Stinner

>                        glibc, clang               https://buildbot.python.org/all/#/builders/435 XXX, XXX

Reviewer: ```suggestion
                       glibc, clang               https://buildbot.python.org/all/#/builders/435 Victor Stinner, XXX
```

s390x-linux-gnu        glibc, gcc                 https://buildbot.python.org/all/#/builders/223 XXX

>                        glibc, clang               https://buildbot.python.org/all/#/builders/3   XXX

Reviewer: Sadly, I don't have access to s390x buildbot workers, so it's harder for me to debug such issue :-( Maybe move it to Tier 3.

> x86_64-linux-gnu       glibc, clang               https://buildbot.python.org/all/#/builders/441 XXX

Reviewer: You can add me: Victor Stinner

> x86_64-unknown-freebsd BSD libc, cc               https://buildbot.python.org/all/#/builders/172 XXX

Reviewer 2: FreeBSD's cc is clang

```suggestion
x86_64-unknown-freebsd BSD libc, clang               https://buildbot.python.org/all/#/builders/172 XXX
```

> x86_64-unknown-freebsd BSD libc, clang            https://buildbot.python.org/all/#/builders/172 XXX

Reviewer: You can add me: Victor Stinner

> x86_64-linux-gnu       glibc, clang               https://buildbot.python.org/all/#/builders/441 Victor Stinner, XXX

Reviewer 5: I'm not clear on how this differs from tier 1 `x86_64-\*-linux-gnu` beyond adding `clang` to the list and a mystery wildcard (what's that for) being removed but feel free to list my name.

Author: It's just clang. The wildcard is a more complete naming that @tiran requested.

====================== ========================== ============================================== ========

All other platforms
-------------------

Support for a platform may be partial within the code base, such as
from active development around platform support or accidentally.
Code changes to platforms not listed in the above tiers may rejected
or removed from the code base without a deprecation process if they
cause a maintenance burden or obstruct general improvements.

Platforms not listed here may be supported by the wider Python
community in some way. If your desired platform is not listed above,
please perform a search online to see if someone is already providing
support in some form.

Notes
-----

Microsoft Windows

> /////////////////

Reviewer 3: PEP 12 (since you've conformed the other headers to it)

```suggestion
'''''''''''''''''
```

supported in CPython, what platforms are currently supported, and
documents past support.

Rationale
=========

Over time, the CPython source code has collected various pieces of
platform-specific code, which, at some point in time, was
considered necessary to use CPython on a specific platform.
Without access to this platform, it is not possible to determine
whether this code is still needed.  As a result, this code may
either break during CPython's evolution, or it may become
unnecessary as the platforms evolve as well.

Allowing these fragments to grow poses the risk of
unmaintainability: without having experts for a large number of
platforms, it is not possible to determine whether a certain
change to the CPython source code will work on all supported
platforms.

To reduce this risk, this PEP specifies what is required for a
platform to be considered supported by CPython as well as providing a
procedure to remove code for platforms with few or no CPython
users.

This PEP also lists what plaforms *are* supported by the CPython
interpreter. This lets people know what platforms are directly
supported by the CPython development team.

Support tiers
=============

> Platform support is broken down into *tiers*. Each tier comes with
> different requirements which lead to different promises being made
> about support.
> 
> To be promoted to a tier, steering council support is required and is
> expected to be driven by team consensus. Demotion to a lower tier
> occurs then the requirements of the current tier are no longer met for
> a platform for an extended period of time based on the judgment of
> the release manager or steering council. For platforms which no longer
> meet the requirements of any tier by b1 of a new feature release, an
> announcement will be made to warn the community of the pending removal
> of support for the platform (e.g. in the b1 announcement). If the
> platform is not brought into line for at least one of the tiers by the
> first release candidate, it will be listed as unsupported in this PEP.

Reviewer 10: ```suggestion
Platform support is broken down into *tiers*. Each tier comes with
different requirements which lead to different promises being made
about support.

To be promoted to a tier, steering council support is required and is
expected to be driven by team consensus. Demotion to a lower tier
occurs when the requirements of the current tier are no longer met for
a platform for an extended period of time based on the judgment of
the release manager or steering council. For platforms that no longer
meet the requirements of any tier by b1 of a new feature release, an
announcement will be made to warn the community of the pending removal
of support for the platform (e.g. in the b1 announcement). If the
platform is not brought into line for at least one of the tiers by the
first release candidate, it will be listed as unsupported in this PEP.
```

> procedure to remove code for platforms with few or no Python

Reviewer 11: The previous line, which I cannot quote, is
"platform to be considered supported by Python as well as providing a"
Should that be CPython?

Author: Yep. I'll fix it shortly.

source tree 3 years after the extended support for the compiler has
ended (but continue to remain available in revision control).

Legacy C Locale

> ///////////////

Reviewer 3: PEP 12 (since you've conformed the other headers to it)

```suggestion
'''''''''''''''
```
